types/jest/index.d.ts

```diff
@@ -572,7 +576,7 @@ declare namespace jest {
          * This ensures that the object contains the desired structure.
          */
         // tslint:disable-next-line: no-unnecessary-generics
-        objectContaining<E = {}>(obj: E): any;
+        objectContaining<E = {}>(obj: Partial<E>): any;
```

Comment:
Why are these being made `Partial`? If I want the properties of the object I'm typing with the generic to be optional, I can use `Partial` myself.

Reply:
I made this change because I can see no reason for `objectContaining` to accept entire object. From [jest docs](https://jestjs.io/docs/en/expect#expectobjectcontainingobject): 

> That is, the expected object is a **subset** of the received object

If you need to check for entire object, why would you use `.toBeCalledWith(expect.objectContaining(value))` instead of just `.toBeCalledWith(value)`?

Reply:
Being a subset doesn't mean it doesn't require properties - Using `Partial` would make sense if `E` was the object being passed in to `expect`, but that's not required.

Consider this:

```
interface CommonAttributes {
  id: string;
  createdAt: string;
  modifiedAt: string;
}

it('contains the common attributes', () => {
  expect(model.attributes).toStrictEqual(
    expect.objectContaining<CommonAttributes>({
      id: expect.any(Number),
      createdAt: expect.any(String),
      modifiedAt: expect.any(String)
    })
  );
});
```

That is testing that `model` is a subset of `CommonAttributes`, which in order to be true requires specific properties.

Reply:
Thanks for the example, it really makes sense. However I think that this is rather unusual case - in my experience most usages are like `expect.objectContaining<Partial<MyInterface>>({...})`, which seems kinda cumbersome. In your case I would use `expect(model.attributes).toMatchObject<CommonAttributes>(...)` - this one is unchanged.

If there are any other examples, I agree that this particular change should be reverted. Additionally, one can argue that even this example can be considered as unnecessary breaking change.

Please tell me what you think.

Reply:
My example is a simplification of a pretty reasonable one - it's about testing duck-typing, which is actually really powerful, since it lets you leverage TypeScript to improve testing that things are of the right subtype (because i.e if I suddenly decide to require all my models need to have a `slug` as well, TS will tell me everywhere I need to update my tests).

[`toMatchObject` is not a direct replacement for `expect.objectContaining`](https://stackoverflow.com/questions/45692456/whats-the-difference-between-tomatchobject-and-objectcontaining):

```
  // objectContaining, with nested object, containing partial props/values
  // FAILS
  expect({ position: { x: 0, y: 0 } }).toEqual(expect.objectContaining({
    position: {
      x: expect.any(Number)
    }
  }));

  // toMatchObject, with nested object, containing partial props/values
  // PASSES
  expect({ position: { x: 0, y: 0 } }).toMatchObject({
    position: {
      x: expect.any(Number)
    }
  });
```

[It's even been mentioned the possibility of removing it](https://github.com/jest-community/eslint-plugin-jest/issues/230#issuecomment-469253911) in favor of using `expect.objectContaining` (afaik the only reason its still there, aside from removing a matcher being a huge breaking change, is that it's a little extra to type `expect.objectContaining` + a matcher).

> Additionally, one can argue that even this example can be considered as unnecessary breaking change.

I'm not sure what you mean by this.

-----

At the very least, I don't think this change should be made in this PR as it's not required for it and should be considered separate.

I understand why this seems like a reasonable change, and agree it'd be nice, but my issue is **this prevents you from ever marking a property as required**, where the gain is just saving ~5 characters by not typing `Partial` yourself.

Reply:
Thank you for your explanation! I'll remove this change and add a test for it 👍